# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/wgpu-core/src/command/transition_resources.rs
# Commit: 3f167f8b3137
# Full Hash: 3f167f8b313749f50f135e189e5854bef7ac0c73
# Author: Erich Gubler <erichdongubler@gmail.com>
# Date: 2025-10-03 18:08 +0000
# Regressor Bug: 1991551
# File Overlap Count: 23
# Overlapping Files: third_party/rust/wgpu-hal/Cargo.toml, supply-chain/audits.toml, third_party/rust/naga/src/back/spv/writer.rs, third_party/rust/naga/src/back/msl/writer.rs, third_party/rust/wgpu-types/src/lib.rs
# Description:
#   Bug 1991551 - build(webgpu): update WGPU to 8c4aebc0c1b75e3f7412dfc180745a0717fc30ec r=webgpu-reviewers,supply-chain-reviewers,nical
# ==============================================================================

diff -r 823439f307a0 -r 3f167f8b3137 third_party/rust/wgpu-core/src/command/transition_resources.rs
--- a/third_party/rust/wgpu-core/src/command/transition_resources.rs	Fri Oct 03 18:01:30 2025 +0000
+++ b/third_party/rust/wgpu-core/src/command/transition_resources.rs	Fri Oct 03 18:08:29 2025 +0000
@@ -1,12 +1,14 @@
+use alloc::{sync::Arc, vec::Vec};
+
 use thiserror::Error;
 use wgt::error::{ErrorType, WebGpuError};
 
 use crate::{
-    command::{CommandEncoder, CommandEncoderError, EncoderStateError},
+    command::{encoder::EncodingState, ArcCommand, CommandEncoder, EncoderStateError},
     device::DeviceError,
     global::Global,
     id::{BufferId, CommandEncoderId, TextureId},
-    resource::{InvalidResourceError, ParentDevice},
+    resource::{Buffer, InvalidResourceError, ParentDevice, Texture},
     track::ResourceUsageCompatibilityError,
 };
 
@@ -24,54 +26,72 @@
         // Lock command encoder for recording
         let cmd_enc = hub.command_encoders.get(command_encoder_id);
         let mut cmd_buf_data = cmd_enc.data.lock();
-        cmd_buf_data.record_with(|cmd_buf_data| -> Result<(), CommandEncoderError> {
-            // Get and lock device
-            let device = &cmd_enc.device;
-            device.check_is_valid()?;
-            let snatch_guard = &device.snatchable_lock.read();
-
-            let mut usage_scope = device.new_usage_scope();
-            let indices = &device.tracker_indices;
-            usage_scope.buffers.set_size(indices.buffers.size());
-            usage_scope.textures.set_size(indices.textures.size());
-
-            // Process buffer transitions
-            for buffer_transition in buffer_transitions {
-                let buffer = hub.buffers.get(buffer_transition.buffer).get()?;
-                buffer.same_device_as(cmd_enc.as_ref())?;
-
-                usage_scope
-                    .buffers
-                    .merge_single(&buffer, buffer_transition.state)?;
-            }
-
-            // Process texture transitions
-            for texture_transition in texture_transitions {
-                let texture = hub.textures.get(texture_transition.texture).get()?;
-                texture.same_device_as(cmd_enc.as_ref())?;
-
-                unsafe {
-                    usage_scope.textures.merge_single(
-                        &texture,
-                        texture_transition.selector,
-                        texture_transition.state,
-                    )
-                }?;
-            }
-
-            // Record any needed barriers based on tracker data
-            let cmd_buf_raw = cmd_buf_data.encoder.open()?;
-            CommandEncoder::insert_barriers_from_scope(
-                cmd_buf_raw,
-                &mut cmd_buf_data.trackers,
-                &usage_scope,
-                snatch_guard,
-            );
-            Ok(())
+        cmd_buf_data.push_with(|| -> Result<_, TransitionResourcesError> {
+            Ok(ArcCommand::TransitionResources {
+                buffer_transitions: buffer_transitions
+                    .map(|t| {
+                        Ok(wgt::BufferTransition {
+                            buffer: self.resolve_buffer_id(t.buffer)?,
+                            state: t.state,
+                        })
+                    })
+                    .collect::<Result<_, TransitionResourcesError>>()?,
+                texture_transitions: texture_transitions
+                    .map(|t| {
+                        Ok(wgt::TextureTransition {
+                            texture: self.resolve_texture_id(t.texture)?,
+                            selector: t.selector,
+                            state: t.state,
+                        })
+                    })
+                    .collect::<Result<_, TransitionResourcesError>>()?,
+            })
         })
     }
 }
 
+pub(crate) fn transition_resources(
+    state: &mut EncodingState,
+    buffer_transitions: Vec<wgt::BufferTransition<Arc<Buffer>>>,
+    texture_transitions: Vec<wgt::TextureTransition<Arc<Texture>>>,
+) -> Result<(), TransitionResourcesError> {
+    let mut usage_scope = state.device.new_usage_scope();
+    let indices = &state.device.tracker_indices;
+    usage_scope.buffers.set_size(indices.buffers.size());
+    usage_scope.textures.set_size(indices.textures.size());
+
+    // Process buffer transitions
+    for buffer_transition in buffer_transitions {
+        buffer_transition.buffer.same_device(state.device)?;
+
+        usage_scope
+            .buffers
+            .merge_single(&buffer_transition.buffer, buffer_transition.state)?;
+    }
+
+    // Process texture transitions
+    for texture_transition in texture_transitions {
+        texture_transition.texture.same_device(state.device)?;
+
+        unsafe {
+            usage_scope.textures.merge_single(
+                &texture_transition.texture,
+                texture_transition.selector,
+                texture_transition.state,
+            )
+        }?;
+    }
+
+    // Record any needed barriers based on tracker data
+    CommandEncoder::insert_barriers_from_scope(
+        state.raw_encoder,
+        state.tracker,
+        &usage_scope,
+        state.snatch_guard,
+    );
+    Ok(())
+}
+
 /// Error encountered while attempting to perform [`Global::command_encoder_transition_resources`].
 #[derive(Clone, Debug, Error)]
 #[non_exhaustive]